### PR TITLE
Presentation: Navigation property editing

### PR DESCRIPTION
**Pull request by saskliutas**
Closes iTwin/itwinjs-backlog#112

Added editor for navigation properties. This editor can be used to edit navigation property values in property grid and enter navigation property value when building instance filter.

### CODE REVIEW

File: ui/components-react/src/components-react/properties/renderers/PropertyView.tsx [hunk 1]
```diff
@@ -112,7 +112,7 @@ export class PropertyView extends React.Component<PropertyViewProps, PropertyVie
           : undefined}
         {this.props.propertyRecord.value.valueFormat === PropertyValueFormat.Primitive
           ? <div className="components-property-record-value">
-            <span>{this.props.valueElementRenderer ? this.props.valueElementRenderer() : this.props.valueElement}</span>
+            {this.props.valueElementRenderer ? this.props.valueElementRenderer() : this.props.valueElement}
```
Comment (roluk) [inline ui/components-react/src/components-react/properties/renderers/PropertyView.tsx:115]: This is a breaking change for user-defined CSS rules.

**Review by roluk — CHANGES_REQUESTED** [review]

File: presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+export interface NavigationPropertyTargetSelectorProps extends PropertyEditorProps {
+  imodel: IModelConnection;
+  getNavigationPropertyInfo: (record: PropertyRecord) => Promise<NavigationPropertyInfo | undefined>;
+}
+
+export const NavigationPropertyTargetSelector = React.forwardRef<NavigationPropertyTargetSelectorAttributes, NavigationPropertyTargetSelectorProps>((props, ref) => {
+  const { imodel, getNavigationPropertyInfo, propertyRecord, onCommit, setFocus } = props;
+  const divRef = React.useRef<HTMLDivElement>(null);
+  const targetsRuleset = useNavigationPropertyTargetsRuleset(getNavigationPropertyInfo, propertyRecord);
+  const loadTargets = useNavigationPropertyTargetsLoader({ imodel, ruleset: targetsRuleset });
```
Comment (grigasp) [inline presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx:36]: We won't get the ruleset if `getNavigationPropertyInfo` is `undefined`, which means we also won't get any load targets. I suppose that means we're going to see an empty select box?

I think we should render a read-only record if it has a value.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx (hunk 2 above)
Comment (saskliutas) [reply presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx:40]: Changed to use default property renderer to render property value if `targetsRuleset` is `undefined`.

File: presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx [hunk 3]
```diff
@@ -47,8 +49,8 @@ export const NavigationPropertyEditorContext = React.createContext<NavigationPro
 export function useNavigationPropertyEditingContextProps(imodel: IModelConnection, dataProvider: IContentDataProvider): NavigationPropertyEditorContextProps {
   return React.useMemo<NavigationPropertyEditorContextProps>(() => ({
     imodel,
-    getNavigationPropertyInfo: async (record) => {
-      const field = await dataProvider.getFieldByPropertyRecord(record);
+    getNavigationPropertyInfo: async (property) => {
+      const field = await dataProvider.getFieldByPropertyRecord(new PropertyRecord({ valueFormat: PropertyValueFormat.Primitive }, property));
```
Comment (grigasp) [inline presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx:53]: The `ContentDataProvider.getFieldByPropertyRecord` takes a `PropertyRecord` and uses only property description from it. Maybe we should also add a `ContentDataProvider.getFieldByPropertyDescription`? That way we wouldn't need to create a fake `PropertyRecord` here.

File: presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx [hunk 4]
```diff
@@ -9,37 +9,41 @@ import {
...
+  const targetsRuleset = useNavigationPropertyTargetsRuleset(getNavigationPropertyInfo, propertyRecord.property);
   const loadTargets = useNavigationPropertyTargetsLoader({ imodel, ruleset: targetsRuleset });
 
-  const [selectedTarget, setSelectedTarget] = React.useState<NavigationPropertyTarget | undefined>();
+  const [selectedTarget, setSelectedTarget] = React.useState(() => getNavigationTargetFromPropertyRecord(propertyRecord));
 
   const onChange = React.useCallback((target?: NavigationPropertyTarget) => {
     setSelectedTarget(target);
-    target && propertyRecord && onCommit && onCommit({ propertyRecord, newValue: getPropertyValue(target) });
+    target && onCommit && onCommit({ propertyRecord, newValue: getPropertyValue(target) });
```
Comment (grigasp) [inline presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx:46]: Looks like we're only going to call `onCommit` if there is a selected value. What about navigation properties that have an optional target, can we handle those?

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx (hunk 4 above)
Comment (saskliutas) [reply presentation/components/src/presentation-components/properties/NavigationPropertyTargetSelector.tsx:46]: It does not allow to select empty target at the moment. I filled an issue to add support for optional targets #4433

File: presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  // istanbul ignore next
+  public get htmlElement() {
+    return this._ref.current?.divElement ?? null;
+  }
+
+  // istanbul ignore next
+  public get hasFocus() {
+    if (!this._ref.current || !document.activeElement)
+      return false;
+    return document.activeElement.contains(this._ref.current.divElement);
```
Comment (roluk) [inline presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx:79]: Isn't this backwards? The active element should be contained by the editor for it to have focus.

File: presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+PropertyEditorManager.registerEditor(StandardTypeNames.Navigation, NavigationPropertyEditor);
+
+/** @alpha */
+export interface NavigationPropertyEditorContextProps {
+  imodel: IModelConnection;
+  getNavigationPropertyInfo: (property: PropertyDescription) => Promise<NavigationPropertyInfo | undefined>;
+}
+
+/** @alpha */
+export const NavigationPropertyEditorContext = React.createContext<NavigationPropertyEditorContextProps | undefined>(undefined);
```
Comment (roluk) [inline presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx:46]: There is no need to treat React context variable as component.

**Review by grigasp — APPROVED** [review]

**Review by roluk — CHANGES_REQUESTED** [review]

**Review by pmconne — APPROVED** [review]

File: presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx (hunk 5 above)
Comment (saskliutas) [reply presentation/components/src/presentation-components/properties/NavigationPropertyEditor.tsx]: Nice catch. Fixed it.

**Review by roluk — APPROVED** [review]

**Review by raplemie — APPROVED** [review]

**Pull request merged**